### ISSUE TITLE
Add `trim_file` command and `trim_file_on_save` configuration option

### ISSUE THREAD

**Issue by poliorcetics** [helix-editor/helix#1481]
#### Describe your feature request

I know there is an `trim_selection` command but using it breaks the current selection, which is a not very helpful. It's also not very practical to do it for each opened file when ending a somewhat long session.

**Comment by poliorcetics** [comment]
This would maybe need an additional config option to not do it for heavy files, since trimming could be quite long on files of 100k lines ? Even if there is no trimming to be done, each line must be checked individually (though that should be parallel once the initial reading is done).

**Comment by poliorcetics** [comment]
Add-on: maybe a default option + a per language one since markdown uses "two spaces at end of line" as code for "insert visible line break here" but I don't know if any other commonly used language that does something like this

**Comment by EpocSquadron** [comment]
It's worth noting that `trim_selection` is intended to literally change the selections to exclude whitespace from their ends. It does not actually remove the whitespace from the document.

Currently, we rely on the LSP implementation to format on document save, which will often include trimming whitespace from the end of lines. If the LSP is not present or doesn't implement formatting, then you are currently stuck doing it manually. There is discussion of what to do about this elsewhere but I'm on mobile, may find it later and comment.

In the meantime, you can select the entire document (`%`), then pipe to a formatter on the command line (`|prettier<enter>` for example).

**Comment by EpocSquadron** [comment]
#1396 would somewhat alleviate this issue.

**Comment by Omnikar** [comment]
> #### Describe your feature request
> 
> I know there is an `trim_selection` command but using it breaks the current selection, which is a not very helpful. It's also not very practical to do it for each opened file when ending a somewhat long session.

As for applying some operation to all the files in a long session, macros can probably help automate that.

**Comment by Omnikar** [comment]
Here's a macro that should trim trailing whitespace from the end of all lines in the current buffer, save it, then close it. If you yank the macro into the `@` register, you should be able to just hold `q` (or type a large enough number followed by `q`) to apply the macro to every open buffer.
```
%s\s+\n<ret>vhvd:w<ret>:bc<ret>
```

**Comment by poliorcetics** [comment]
> #1396 would somewhat alleviate this issue.

I don't like this solution much; it forces users to have a default language server on every new install of helix and to wait for it to spin up to quickly trim a file that failed an `editor-config` check.

**Comment by Paveloom** [comment]
Using the current version from the  `master` branch, you can do the following:

```toml
# ~/.config/helix/config.toml
# <...>
[keys.normal]
C-s = [
    "save_selection",
    "select_all",
    # Delete trailing whitespace from end of each line
    ":pipe sed 's/[ \t]*$//'",
    # Delete trailing blank lines (including whitespace) from end of the buffer
    ":pipe awk '/^\\s*$/ {b=b $0 \"\\n\"; next;} {printf \"\\%s\",b; b=\"\"; print;}'",
    "collapse_selection",
    "jump_backward",
    ":write",
]
# <...>
```

Current issues:
- the `":write"` command writes but also introduces a phantom change;
- if you left the cursor in the targeted whitespace area, the program may crash on the backward jump.

**Comment by vaxXxa** [comment]
Would be nice to have similar to VSCode flags:

<img width="638" alt="image" src="https://user-images.githubusercontent.com/139175/203183955-847a7aa4-8e50-453a-be87-a86c4c95a5ef.png">
<img width="459" alt="image" src="https://user-images.githubusercontent.com/139175/203183984-4c453d3f-7d49-41bb-80e1-1568290ca283.png">

**Comment by rcorre** [comment]
Note that if you have `git` installed, you can use `git stripspace` as a language formatter:

```
[[language]]
name = "gdscript"
language-server = { command = "nc", args = ["localhost", "6005"], language-id = "gdscript" }
formatter = { command = "git", args = ["stripspace"] }
```

**Comment by DannyJJK** [comment]
> Note that if you have `git` installed, you can use `git stripspace` as a language formatter:
> 
> ```
> [[language]]
> name = "gdscript"
> language-server = { command = "nc", args = ["localhost", "6005"], language-id = "gdscript" }
> formatter = { command = "git", args = ["stripspace"] }
> ```

I'm not sure if that's a good idea, the documentation for this says:
> NOTE: This is intended for cleaning metadata, prefer the --whitespace=fix mode of [git-apply[1]](https://git-scm.com/docs/git-apply) for correcting whitespace of patches or files in the repository.

**Comment by Kurt-Shiwz** [comment]
> Using the current version from the `master` branch, you can do the following:
> 
> ```toml
> # ~/.config/helix/config.toml
> # <...>
> [keys.normal]
> C-s = [
>     "save_selection",
>     "select_all",
>     # Delete trailing whitespace from end of each line
>     ":pipe sed 's/[ \t]*$//'",
>     # Delete trailing blank lines (including whitespace) from end of the buffer
>     ":pipe awk '/^\\s*$/ {b=b $0 \"\\n\"; next;} {printf \"%s\",b; b=\"\"; print;}'",
>     "collapse_selection",
>     "jump_backward",
>     "commit_undo_checkpoint",
>     ":write",
> ]
> # <...>
> ```
> 
> Current issues:
> 
> * the `commit_undo_checkpoint` command is a workaround ([Using `:write` after any command that produces changes introduces a phantom change #2192](https://github.com/helix-editor/helix/issues/2192));
> * if you left the cursor in the targeted whitespace area, the program may crash on the backward jump ([Occasional crash when jumping to previous location #1238](https://github.com/helix-editor/helix/issues/1238)).
> 
> References:
> 
> * https://stackoverflow.com/a/61516740
> * http://sed.sourceforge.net/sed1line.txt

thx, i try to keep the file has new line on end of file, it work.
```
C-s = [
    "save_selection",
    "select_all",
    # Delete trailing whitespace from end of each line
    ":pipe sed 's/[ \t]*$//'",
    # Delete trailing blank lines (including whitespace) from end of the buffer
    ":pipe awk '/^\\s*$/ {b=b $0 \"\\n\"; next;} {printf \"%s\",b; b=\"\"; print;}'",
    "collapse_selection",
    "goto_last_line",
    "add_newline_below",
    "commit_undo_checkpoint",
    ":write",
]
```

**Comment by rcorre** [comment]
FWIW, if you use `sed` as a formatter in `languages.toml`, you don't need all the selection/checkpoint stuff:

```formatter = { command = "sed", args = ["s/[ \t]*$//"] }
```

**Comment by milanglacier** [comment]
FYI, if you don't want to exit file after trimming white spaces, but you want to keep editing the files, you may find the followin command to be helpful in order to keep the original cursor position:


```toml
"space" = { "m" = [ "save_selection", "select_all", ":pipe sed 's/[[:space:]]*$//'", "collapse_selection", "jump_backward", "flip_selections" ] }
```

The trick here is you need to use `jump_backward` with `flip_selections` together to restore the original cursor position.

**Comment by jficz** [comment]
> The trick here is you need to use `jump_backward` with `flip_selections` together to restore the original cursor position.

does that really work for you? On 24.07 when the selection is piped, the jumplist is cleared

**Comment by milanglacier** [comment]
> > The trick here is you need to use `jump_backward` with `flip_selections` together to restore the original cursor position.
> 
> does that really work for you? On 24.07 when the selection is piped, the jumplist is cleared

yes, this no longer works for me. updated to 25.1 and found that my trick no longer works.

**Comment by jficz** [comment]
Might that be a regression? I mean technically it makes sense because when the selection is piped and replaced, Helix has no way of knowing what the result will be and the jump marks might not be applicable anymore. On the other hand, clearing the jumplist outright is also a bit drastic. Not sure if it makes sense to open an issue or not.

**Comment by the-mikedavis** [comment]
I've added configuration options matching those VSCode settings in aa20eb8e7f63eaab8af061782ae32e0dcea5c73e

```toml
[editor]
# Trim whitespace preceding line endings on write (default false)
trim-trailing-whitespace = true
# Trim extra line endings after the final one on write (default false)
trim-final-newlines = true
```

I think it's still worthwhile to have a command for this. (Probably it should work on the line ranges of the selection)

**Comment by andradei** [comment]
Can't wait for this to land.

**Comment by milanglacier** [comment]
> Most notably, markdown creates a linebreak (but not a new paragraph) on two trailing spaces in a line.

While I know this is a special feature. But I want to suppress this feature., that is I also want to automatically trim white spaces on markdown file.

This special feature is also documented in the [linebreak best practices](https://www.markdownguide.org/basic-syntax/#line-break-best-practices) section from markdownguide.org as controversial.

With that being said, I think `trim-whitespaces` setting should be added to part of `languages.toml` and can be set per language.

So user who can override the global trim-whitespaces settings per language.

**Comment by Chasarr** [comment]
Should this issue be closed perhaps? It seems to be solved

**Comment by the-mikedavis** [comment]
I'd still like to see typable commands that apply the trimming to the current selection(s) before calling this closed
